### PR TITLE
Group hierarchy improvements (by 3298)

### LINKED ISSUE #5977: Vehicle Groups: improve error message when group hierarchy fails
**_juanjo_ opened the ticket and wrote:**

> In the recent feature about group hierarchy, the error message doesn't explain that loops in group hierarchy cannot be created.
> 
> I think the error message could explain why that action cannot be done, as some of the error messages do: "That action can't be done... Reason is that..."

**Reported version**: _trunk_
**Operating system**: _All_

---

### LINKED ISSUE #6053: Vehicle Groups: packing and unpacking subgroups
**_Milsa_ opened the ticket and wrote:**

> Please add pack and unpack subgroups functionality to group view.

**Reported version**: _Version?_
**Operating system**: _All_

---

### LINKED ISSUE #6189: Vehicle Groups: parent group should count total number of vehicles in subgroups
**_2006TTD_ opened the ticket and wrote:**

> For ease of discussion I attached a screenshot showing a sample multi-level vehicle group.
> 
> First, I understand that in "Level 2" the grayed out entries are those that belong to the subgroups, this has no problem.
> 
> However, the number of vehicles shown on the left panel (i.e. the column at which the yellow arrow is pointing) may be confusing. 
> "Level 2" is displaying the figure of 1 - one vehicle which is in the group, but not in any of the subgroups, but some may be thinking it should be displaying 1+1+3+4+0 = 9
> 
> A similar problem is also seen in "2A", it may be debatable on whether it should be 1 or 1+3+4 = 8
> 
> To tackle this problem, there may be three solutions:
> (1) Leave it as it is and add to the documentation that the numbers are meant for the vehicles that belong to that group at that level exclusively, but that is not very reasonable as many would expect a sum of the group and the subgroups from a mathematical perspective.
> 
> (2) Display the total number of vehicles in that group and subgroups
> i.e. Level 1 would show 0+2+1+2 = 5; Level 2 showing 9; 2A showing 8 and so on.
> 
> (3) Display both figures in the form
> (Number of vehicles in group only) / (Total number of vehicles in group and subgroups)
> i.e. Level 1 would show 0/5; Level 2 showing 1/9; 2A showing 1/8 and so on.
> 
> 

**Attachments**
- [Groups.png](https://bugs.openttd.org/task/6189/getfile/10057/Groups.png) (75.98 KiB)

**Reported version**: _trunk_
**Operating system**: _All_

---

### PR DESCRIPTION
**Pull request by LordAro**
From https://www.tt-forums.net/viewtopic.php?f=33&t=83025

Group patches 1 & 2 have been previously implemented, so this is 3-5

Fix #5977
Fix #6053 
Fix #6189

### CODE REVIEW

File: src/group_gui.cpp [hunk 1]
```diff
@@ -253,6 +271,12 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
 		const GroupStatistics &stats = GroupStatistics::Get(this->vli.company, g_id, this->vli.vtype);
 		bool rtl = _current_text_dir == TD_RTL;
 
+		/* draw fold / unfold button */
+		int x = rtl ? right - WD_FRAMERECT_RIGHT - 8 - this->column_size[VGC_FOLD].width + 1 : left + WD_FRAMERECT_LEFT + 8;
+		if (has_children) {
+			DrawSprite(Group::Get(g_id)->folded ? SPR_CIRCLE_FOLDED : SPR_CIRCLE_UNFOLDED, PAL_NONE, rtl ? x - indent * LEVEL_WIDTH : x + indent * LEVEL_WIDTH, y + (this->tiny_step_height - this->column_size[VGC_FOLD].height) / 2);
```
Comment (PeterN) [inline src/group_gui.cpp]: indent level has already been multiplied by LEVEL_WIDTH by the caller of DrawGroupInfo, hence the display issue.

File: src/group_gui.cpp [hunk 2]
```diff
@@ -672,6 +696,33 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
 				uint id_g = this->group_sb->GetScrolledRowFromWidget(pt.y, this, WID_GL_LIST_GROUP, 0, this->tiny_step_height);
 				if (id_g >= this->groups.Length()) return;
 
+				if (groups[id_g]->folded || (id_g + 1 < this->groups.Length() && this->indents[id_g + 1] > this->indents[id_g])) {
+					/* The group has children, check if the user clicked the fold / unfold button. */
+					NWidgetCore *group_display = this->GetWidget<NWidgetCore>(widget);
+					int x = _current_text_dir == TD_RTL ?
+							group_display->pos_x + group_display->current_x - WD_FRAMERECT_RIGHT - 8 - this->indents[id_g] * LEVEL_WIDTH - this->column_size[VGC_FOLD].width :
+							group_display->pos_x + WD_FRAMERECT_LEFT + 8 + this->indents[id_g] * LEVEL_WIDTH;
+					if (click_count > 1 || (pt.x >= x && pt.x < (int) (x + this->column_size[VGC_FOLD].width))) {
```
Comment (PeterN) [inline src/group_gui.cpp]: `(int)(...)`

File: src/group_gui.cpp [hunk 3]
```diff
@@ -672,6 +696,33 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
...
+
+						GroupID g = this->vli.index;
+						if (!(IsAllGroupID(g) || IsDefaultGroupID(g))) {
```
Comment (PeterN) [inline src/group_gui.cpp]: `if (!IsAllGroupID(g) && !IsDefaultGroupID(g)) {` perhaps?

File: src/group_cmd.cpp [hunk 4]
```diff
@@ -298,6 +298,7 @@ void PropagateChildLivery(const Group *g)
 Group::Group(Owner owner)
 {
 	this->owner = owner;
+	this->folded = false;
```
Comment (PeterN) [inline src/group_cmd.cpp:301]: Is this not `false` by default?

**Review by PeterN — DISMISSED** [review]

File: src/group_cmd.cpp (hunk 4 above)
Comment (LordAro) [reply src/group_cmd.cpp:301]: i always get confused about when members are zeroed on initialisation or not. No harm in being explicit?

Comment (PeterN) [reply src/group_cmd.cpp:301]: No harm.

**Comment by LordAro** [comment]
image
fixed.

**Review by PeterN — DISMISSED** [review]

**Review by PeterN — APPROVED** [review]

**Pull request merged**